### PR TITLE
feat(cli): size purchases by target RI/SP utilization (--target-utilization, closes #338)

### LINKED ISSUE #338: feat(cli): size purchases by target RI/SP utilization (--target-utilization)
## Summary

Add a CLI sizing mode that picks commitment purchases (RIs / Savings Plans) to keep **projected RI / SP utilization** at or above a user-specified target — instead of (or alongside) the current `--coverage` percentage of recommendations.

## Motivation

Today the CLI sizes purchases via `--coverage` (default 80%), which means *"buy enough to cover N% of recommended usage."* That answers "how much demand do we cover?" but ignores the dual question: *"how much of what we buy will actually be used?"*

Over-purchasing in a workload with bursty / seasonal traffic produces commitments with low utilization — wasted spend on the very tool meant to reduce spend. A utilization-targeted mode lets users say *"buy commitments only as long as projected utilization stays ≥ 95%"*, which is a more conservative, ROI-protective stance — particularly useful for accounts that already carry meaningful existing commitments.

## Proposed behavior

Add a new flag, mutually-exclusive-ish with `--coverage`:

```
--target-utilization, -u   float64   Target post-purchase utilization % (0-100).
                                      When set, sizes recommendations so projected
                                      utilization stays >= target. Overrides
                                      --coverage when both are provided.
```

Sizing algorithm (high level — pseudocode, not committing to specifics):

1. For each instance-family + region + tenancy (or SP equivalent):
   - Fetch existing-commitment **utilization** from Cost Explorer
     `GetReservationUtilization` / `GetSavingsPlansUtilization` over the same
     lookback window the rest of CUDly uses.
   - Fetch existing-commitment **count** (active normalized units).
   - From the existing recommendation engine, get the upper-bound purchase
     suggestion (this is what `--coverage 100` would buy).
2. Walk the recommendation from 0 → upper-bound in increments. At each step:
   - Project post-purchase utilization assuming the historical usage curve.
   - Stop at the largest quantity that still keeps projected utilization
     >= `--target-utilization`.
3. Emit the chosen quantity with both metrics in the output: projected
   utilization, projected coverage. (`ri-helper-dryrun-*.csv` already
   carries the per-row data — add two columns.)

## Open questions for triage / design

- **Signal source**: `GetReservationUtilization` (Cost Explorer) is the
  canonical source, but it's lagged ~24h and rounded to days. Is that
  fresh enough, or do we need CloudWatch `EC2 - RI Utilization` metrics
  for tighter lookback? See `ec2-rightsizing/` for the existing
  CloudWatch-metric patterns in this repo — likely reusable.
- **SP vs RI**: SPs are dollar-denominated; utilization is computed on
  $/hr commitment used vs purchased. The math is slightly different but
  the algorithm structure is the same.
- **Lookback window**: reuse `--lookback` if it exists, or introduce
  `--utilization-lookback` (default 30d to match CE's UI default).
- **Interaction with `--coverage`**: I lean toward "if `--target-utilization`
  is set, ignore `--coverage`; if neither, default to `--coverage 80`
  (current behavior)." Open to "AND" semantics — pick the lower of the
  two quantities — if that turns out to be what people actually want.

## Out of scope (file separate issues if needed)

- Multi-account rollup of utilization (today's coverage flag is per-run;
  same scoping applies here).
- Surface utilization in the frontend recommendations table (separate
  frontend issue if desired).

## Acceptance criteria

- [ ] New `--target-utilization` flag plumbed through `cmd/main.go` →
      `multi_service.go` recommendation filter.
- [ ] Utilization fetched via Cost Explorer `GetReservationUtilization` /
      `GetSavingsPlansUtilization` (cached per-run alongside the existing
      CloudWatch metric cache).
- [ ] Dry-run CSV (`ri-helper-dryrun-*.csv`) includes `projected_utilization`
      and `projected_coverage` columns.
- [ ] Tests in `cmd/multi_service_test.go` cover: target met at lower
      quantity than coverage default; target unreachable → 0 purchases
      with explanatory log line; target == 100 → only purchases with
      perfectly-matched existing usage; mutual-exclusion behavior with
      `--coverage`.
- [ ] README / `--help` text explains the trade-off vs `--coverage`.

---

## Triage notes (added by @cmagherusan on 2026-05-12)

Codebase exploration surfaced three corrections to the original framing:

**1. `--coverage` semantics — clarification.** The body above says coverage
   = "% of recommendations to purchase." `cmd/helpers.go:87` (`ApplyCoverage`)
   actually scales the **instance count of each recommendation**:
   `newCount = int(float64(rec.Count) * coverage / 100)`. So `--coverage 80`
   on a 10-instance recommendation buys 8 of those instances — it does not
   filter the list of recommendations. The design intent below survives,
   but the mental model should be "scale each rec's count" not "drop recs".

**2. No separate CE API call is needed — data is already in the existing
   recommendation response.** `ReservationPurchaseRecommendationDetail`
   (AWS SDK) already exposes `AverageNumberOfInstancesUsedPerHour`,
   `MaximumNumberOfInstancesUsedPerHour`, and `AverageUtilization`.
   `SavingsPlansPurchaseRecommendationDetail` exposes
   `EstimatedAverageUtilization`. These fields are not currently parsed by
   `providers/aws/recommendations/client.go` (`parseRecommendationDetail`),
   but they're free — already on the wire. The "Option (a)" path in the
   issue's open questions is the right one; the `GetReservationUtilization`
   detour is unnecessary.

**3. Semantics inversion — higher target ⇒ fewer instances.** Intuition
   reads "target utilization 90%" as "buy more to hit 90%", but for
   commitments it's the opposite: utilization = `used / bought`, so
   raising the target *reduces* the recommended buy quantity. For RIs:
   `N_target = ceil(avg_used_per_hour / (target/100))`, capped above by
   the AWS-recommended quantity. The flag is fundamentally a **waste
   ceiling** — "don't buy more than this fraction of waste tolerance."
   The `--help` text and README must spell this out, or users will set
   `--target-utilization 50` expecting to be aggressive and get more
   commitments than `--target-utilization 95`.

**4. Mutual exclusion with `--coverage`.** The current pipeline applies
   `applyCommonCoverage` unconditionally (`cmd/multi_service.go:624`).
   Decision: when `--target-utilization` is set, skip the `--coverage`
   reduction entirely and ignore the `--coverage` flag value (logging an
   info line if both are explicit). Cleaner than min-of-two semantics
   and matches the conceptual framing ("two different sizing modes").

**5. Savings Plans branch.** `HourlyCommitment` is the sizing lever, not
   instance count. CE returns `EstimatedAverageUtilization` for the full
   recommendation. Sizing: when AWS-projected utilization is already at
   or above target, no scaling; when below, scale `HourlyCommitment` down
   by `aws_projected / target` so projected utilization rises to the
   target. (The math assumes the on-demand demand curve is roughly
   linear in commitment $ — a defensible first-cut approximation.)

These corrections are baked into the implementation plan; this comment
preserves the original framing for posterity.

### PR DESCRIPTION
**Pull request by cristim**
## Summary

- Adds a new CLI flag `--target-utilization` (`-u`) that sizes RI/SP purchases by **projected post-purchase utilization** instead of `--coverage` % of recommendations.
- Reuses utilization signals already on the wire (`AverageNumberOfInstancesUsedPerHour`, `AverageUtilization`, `EstimatedAverageUtilization`) — **no new AWS API call**.
- Emits three new columns on the dry-run/purchase CSV: `ProjectedUtilization`, `ProjectedCoverage`, `RecommendedUtilization`.

Closes #338.

## Why a separate sizing mode

`--coverage 80` is "buy 80% of each recommendation's instance count." It answers "how much of demand do we cover?" but ignores "how much of what we buy will actually be used?" In bursty/seasonal workloads, `--coverage 80` can still over-buy, leaving commitments idle.

`--target-utilization 95` is "buy enough that projected utilization stays ≥ 95%." Counter-intuitively, **higher target ⇒ fewer commitments bought** (the math is `used÷bought`; raising the target shrinks the bought side). Use it when under-buying-and-spilling-to-on-demand beats over-buying-and-wasting-commitment.

The two modes are mutually exclusive: when `--target-utilization` is set, `--coverage` is ignored with an info log.

## Math

- **RI**: `n_target = floor(AverageInstancesUsedPerHour / (target/100))`, capped at `rec.Count` (never exceed AWS's recommended ceiling). Drop with explanatory INFO log when `n_target == 0` (target unreachable). Pass through unmodified when avg is zero (no signal); counted in a single end-of-run skip summary.
- **SP**: if `RecommendedUtilization >= target`, no scaling (AWS already projects above target). Else `ratio = RecommendedUtilization / target` (always < 1); scales `SavingsPlanDetails.HourlyCommitment` and `rec.EstimatedSavings` — **exactly the fields `ApplyCoverage` scales on the SP branch**, so the two modes stay structurally consistent. Pass through when `RecommendedUtilization <= 0` OR `HourlyCommitment <= 0` (CE placeholder rows).

## Changes

| Commit | Scope |
|---|---|
| `7d7baf6d1` | `pkg/common`: 4 new fields on Recommendation |
| `bbbc36a24` | parser_ri.go + parser_sp.go: populate the new fields |
| `7d83ec948` | `cmd/main.go`: flag + Config field + validation |
| `79cf1a5ad` | `cmd/helpers.go`: ApplyTargetUtilization + applySizing |
| `9468b0aaf` | `cmd/multi_service*.go`: route both call sites through applySizing |
| `5f1e982e1` | CSV writer: 3 new columns |
| `d40001f6a` | README: flag table + semantics explainer |
| `2bded0b15` | Post-impl review: SP edge cases + target=100 test |

## Test plan

- [x] `go build ./...` clean
- [x] `go test ./...` — full suite passes (no regressions in 30+ existing packages)
- [x] `cudly --help` shows `--target-utilization` with the inversion-warning copy
- [x] `TestApplyTargetUtilization_RI` covers happy path / AWS-ceiling cap / target unreachable drop / no signal pass-through / target met exactly
- [x] `TestApplyTargetUtilization_RI_Target100` covers the AC's "target == 100 → only purchases with perfectly-matched existing usage" boundary
- [x] `TestApplyTargetUtilization_RI_CostScaling` — cost fields untouched on RI branch (matches ApplyCoverage)
- [x] `TestApplyTargetUtilization_SP` — no-op / scale-down / no-signal
- [x] `TestApplyTargetUtilization_SP_NoSignalGuards` — `HourlyCommitment=0` + wrong-Details-type edge cases
- [x] `TestApplySizing` — routes based on `cfg.TargetUtilization > 0`
- [x] `TestValidateTargetUtilization` — range + both-flags-set behaviour
- [x] `TestParseRIUtilizationSignals` + `TestParseSavingsPlanDetail_RecommendedUtilization` — fall back to zero on nil/unparseable
- [x] `TestWriteMultiServiceCSVReport_UtilizationColumns` — header + blank-when-zero formatting

## Design notes

- Two self-review passes (plan: 10 passes against the design; impl: a fresh-eyes Sonnet fork against the actual code) caught: floor-vs-ceil math, missing `ProjectedCoverage` AC field, wrong parser function for SPs (`parseSavingsPlanDetail`, not `parseRecommendationDetail`), wrong enum identifier names, wrong SP cost-field list, missing explanatory log for unreachable target, and two SP edge cases in this branch. All fixed before the push.
- Frontend visibility of the new columns is **out of scope** for this PR per the issue's "Out of scope" section. The CSV change is for the CLI's existing CSV output, not the React frontend.

## Summary by CodeRabbit

* **New Features**
  * Added --target-coverage sizing (supersedes --coverage when set), plus --rebuy-window-days and --min-pool-size flags.
  * Fetches existing RI coverage, accounts for expiring commitments, and adds RDS family-NU sizing to hit target coverage.
  * CSV report expanded with utilization/coverage columns, richer RDS fields, sorting and a TOTAL row.
* **Documentation**
  * README clarifies --coverage vs --target-coverage semantics and precedence.
* **Tests**
  * Large suite of unit tests added/extended for sizing, coverage, CSV and validation logic.
* **Other**
  * Summary output now shows estimated savings as monthly.

[Review Change Stack](https://app.coderabbit.ai/change-stack/LeanerCloud/CUDly/pull/339)

### CODE REVIEW

**Comment by cristim** [comment]
@coderabbitai review

**Comment by cristim** [comment]
@coderabbitai review

**Comment by cristim** [comment]
@coderabbitai review

**Comment by cristim** [comment]
@coderabbitai review

File: providers/aws/recommendations/coverage.go [hunk 1]
```diff
@@ -0,0 +1,475 @@
...
+			}
+		}
+		for _, engine := range rdsCoverageEngines {
+			if err := c.fetchCoverageForRDSEngineRegion(ctx, startStr, endStr, windowHours, engine, region, out); err != nil {
+				return nil, fmt.Errorf("fetching coverage for RDS engine %q region %q: %w", engine, region, err)
+			}
+		}
+	}
+	return out, nil
+}
```
[the remark this answers is not in the document]
Comment (cristim) [reply providers/aws/recommendations/coverage.go:205]: Thanks for flagging — this finding is based on a misread of the current code; politely declining the suggested revert.

**The flag is `--target-coverage`, not `--target-utilization`.** The PR renamed it during the design iteration because the semantics changed: "utilization" (how full each RI is) is always ~100% under the new sizing, while "coverage" (what % of demand is covered) is the user-facing knob. CE `GetReservationCoverage` is the *canonical* source for that signal — the AWS console's reservations-coverage report comes from the same API. There's no way to compute existing-pool coverage from the rec API payload alone (it returns per-account-curated recs, not org-wide pool aggregates).

**The CE coverage fetch is already opt-in.** `cmd/multi_service.go:40`:

```go
func fetchExistingCoverage(...) recommendations.PoolCoverageMap {
    if cfg.TargetCoverage <= 0 {
        return nil
    }
    ...
}
```

Users on the legacy `--coverage` path pay no CE cost and need no new IAM. `ce:GetReservationCoverage` is required only when `--target-coverage` is set, which is documented in the README and surfaces as an opt-in feature gate.

**Why the legacy `--coverage` path can't reuse the new approach.** `--coverage` operates on AWS rec.Count directly (scale all counts by `coverage/100`). It doesn't subtract existing coverage and was never meant to — it's the simpler "buy X% of what AWS suggests" knob. The new `--target-coverage` is a different semantic: "size so the pool reaches Y% covered including existing RIs", which structurally requires the coverage signal.

Closing this thread as dismissed with justification.

**Comment by cristim** [comment]
@coderabbitai review
